src/hotspot/os/posix/malloctrace/mallocTracePosix.cpp

```diff
@@ -0,0 +1,2652 @@
+              copies[pos]._count = entry->count();
+
+              total_size += entry->size();
+              total_count += entry->count();
+
+              pos += 1;
+            }
+
+            entry = entry->next();
+          }
+        }
+
+        lockedTime.stop();
+        assert(pos <= expected_size, "Size must be correct");
+      } else {
+        nr_of_entries[idx] = 0;
+        failed_alloc = true;
+        lockedTime.stop();
+        continue;
+      }
+    }
+
+    // See if it makes sense to trim. We have to shave of enough and don't
+    // trim anyway after sorting.
+    if ((pos < expected_size - 16) && (pos < max_entries)) {
+      void* result = real_malloc_funcs->realloc(entries[idx], pos * sizeof(StatEntryCopy));
+
+      if (result != NULL) {
+        entries[idx] = (StatEntryCopy*) result;
+      }
+    }
+
+    nr_of_entries[idx] = pos;
+    total_entries += expected_size;
+    total_non_empty_entries += pos;
+
+    // Now sort so we might be able to trim the array to only contain the
+    // maximum possible entries.
+    if (spec._sort_by_count) {
+        qsort(entries[idx], nr_of_entries[idx], sizeof(StatEntryCopy), sort_by_count);
+    } else {
+        qsort(entries[idx], nr_of_entries[idx], sizeof(StatEntryCopy), sort_by_size);
+    }
+
+    // Free up some memory if possible.
+    if (nr_of_entries[idx] > max_entries) {
+      void* result = real_malloc_funcs->realloc(entries[idx], max_entries * sizeof(StatEntryCopy));
+
+      if (result == NULL)  {
+        // No problem, since the original memory is still there. Should not happen
+        // in reality.
+      } else {
+        entries[idx] = (StatEntryCopy*) result;
+      }
+
+      nr_of_entries[idx] = max_entries;
+    }
+  }
+
+  uint64_t size_limit = total_size;
+  uint64_t count_limit = total_count;
+
+  if (spec._dump_percentage > 0) {
+    if (spec._sort_by_count) {
+      count_limit = (uint64_t) (0.01 * total_count *  spec._dump_percentage);
+    } else {
+      size_limit = (uint64_t) (0.01 * total_size *  spec._dump_percentage);
+    }
+  }
+
+  AddressHashSet filter_cache(!spec._on_error);
+
+  int curr_pos[NR_OF_STACK_MAPS];
+  memset(curr_pos, 0, NR_OF_STACK_MAPS * sizeof(int));
+
+  uint64_t printed_size = 0;
+  uint64_t printed_count = 0;
+  int printed_entries = 0;
+
+  for (int i = 0; i < max_entries; ++i) {
+    int max_pos = -1;
+    StatEntryCopy* max = NULL;
+
+    // Find the largest entry not currently printed.
+    if (spec._sort_by_count) {
+      for (int j = 0; j < NR_OF_STACK_MAPS; ++j) {
+        if (curr_pos[j] < nr_of_entries[j]) {
+          if ((max == NULL) || (max->_count < entries[j][curr_pos[j]]._count)) {
+            max = &entries[j][curr_pos[j]];
+            max_pos = j;
+          }
+        }
+      }
+    } else {
+      for (int j = 0; j < NR_OF_STACK_MAPS; ++j) {
+        if (curr_pos[j] < nr_of_entries[j]) {
+          if ((max == NULL) || (max->_size < entries[j][curr_pos[j]]._size)) {
```

Comment:
I see now that the dereference of `entries[j]` is not reached because `nr_of_entries[j] == 0` if `entries[j] == NULL`.
I wonder if static code analysis is smart enough ;)